frame/proxy/src/lib.rs

```diff
@@ -0,0 +1,273 @@
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Proxy {
+		/// The set of account proxies. Maps the account which has delegated to the accounts
+		/// which are being delegated to, together with the amount held on deposit.
+		pub Proxies: map hasher(twox_64_concat) T::AccountId => (Vec<(T::AccountId, T::ProxyType)>, BalanceOf<T>);
+	}
+}
+
+decl_error! {
+	pub enum Error for Module<T: Trait> {
+		/// There are too many proxies registered.
+		TooMany,
+		/// Proxy registration not found.
+		NotFound,
+		/// Sender is not a proxy of the account to be proxied.
+		NotProxy,
+		/// A call with a `false` `IsCallable` filter was attempted.
+		Uncallable,
+		/// A call which is incompatible with the proxy type's filter was attempted.
+		Unproxyable,
+		/// Account is already a proxy.
+		Duplicate,
+		/// Call may not be made by proxy because it may escalate its privileges.
+		NoPermission,
+	}
+}
+
+decl_event! {
+	/// Events type.
+	pub enum Event {
+		/// A proxy was executed correctly, with the given result.
+		ProxyExecuted(DispatchResult),
+		/// A call with a `false` IsCallable filter was attempted.
+		Uncallable(u32),
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		/// Deposit one of this module's events by using the default implementation.
+		fn deposit_event() = default;
+
+		/// Dispatch the given `call` from an account that the sender is authorised for through
+		/// `add_proxy`.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// Parameters:
+		/// - `real`: The account that the proxy will make a call on behalf of.
+		/// - `force_proxy_type`: Specify the exact proxy type to be used and checked for this call.
+		/// - `call`: The call to be made by the `real` account.
+		///
+		/// # <weight>
+		/// P is the number of proxies the user has
+		/// - Base weight: 19.87 + .141 * P µs
+		/// - DB weight: 1 storage read.
+		/// - Plus the weight of the `call`
+		/// # </weight>
+		#[weight = {
+			let di = call.get_dispatch_info();
+			(T::DbWeight::get().reads(1)
+				.saturating_add(di.weight)
+				.saturating_add(20 * WEIGHT_PER_MICROS)
+				.saturating_add((140 * WEIGHT_PER_NANOS).saturating_mul(T::MaxProxies::get().into())),
+			di.class)
+		}]
+		fn proxy(origin,
+			real: T::AccountId,
+			force_proxy_type: Option<T::ProxyType>,
+			call: Box<<T as Trait>::Call>
+		) {
+			let who = ensure_signed(origin)?;
+			ensure!(T::IsCallable::filter(&call), Error::<T>::Uncallable);
+			let (_, proxy_type) = Proxies::<T>::get(&real).0.into_iter()
+				.find(|x| &x.0 == &who && force_proxy_type.as_ref().map_or(true, |y| &x.1 == y))
+				.ok_or(Error::<T>::NotProxy)?;
+			match call.is_sub_type() {
+				Some(Call::add_proxy(_, ref pt)) | Some(Call::remove_proxy(_, ref pt)) =>
+					ensure!(&proxy_type == pt, Error::<T>::NoPermission),
+				_ => (),
+			}
+			ensure!(proxy_type.filter(&call), Error::<T>::Unproxyable);
```

Comment:
I'm concerned with utility.batch as a way to bypass this filter.

Reply:
proxy filters are different to call filters, call filters are enforced throughout, including in utility.